File: java_implementation/Scheduler.java
import java.util.*;

public class Scheduler {
    static final String[] DAYS = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
    static final String[] SHIFTS = {"morning", "afternoon", "evening"};

    Map<String, Map<String, List<String>>> preferences = new HashMap<>();
    Map<String, Map<String, List<String>>> schedule = new HashMap<>();
    Map<String, Integer> daysWorked = new HashMap<>();

    public Scheduler() {
        for (String day : DAYS) {
            schedule.put(day, new HashMap<>());
            for (String shift : SHIFTS) {
                schedule.get(day).put(shift, new ArrayList<>());
            }
        }
    }

    public void addEmployeePreference(String name, String day, String shift) {
        preferences.putIfAbsent(name, new HashMap<>());
        preferences.get(name).putIfAbsent(day, new ArrayList<>());
        preferences.get(name).get(day).add(shift);
    }

    public void assignShifts() {
        Random rand = new Random();

        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                List<String> candidates = new ArrayList<>();

                for (String emp : preferences.keySet()) {
                    if (preferences.get(emp).getOrDefault(day, new ArrayList<>()).contains(shift)
                            && daysWorked.getOrDefault(emp, 0) < 5
                            && !alreadyAssigned(day, emp)) {
                        candidates.add(emp);
                    }
                }

                Collections.shuffle(candidates);
                List<String> assigned = candidates.subList(0, Math.min(2, candidates.size()));
                for (String emp : assigned) {
                    schedule.get(day).get(shift).add(emp);
                    daysWorked.put(emp, daysWorked.getOrDefault(emp, 0) + 1);
                }

                while (schedule.get(day).get(shift).size() < 2) {
                    List<String> available = new ArrayList<>();
                    for (String emp : preferences.keySet()) {
                        if (daysWorked.getOrDefault(emp, 0) < 5 && !alreadyAssigned(day, emp)) {
                            available.add(emp);
                        }
                    }
                    if (available.isEmpty()) break;
                    String pick = available.get(rand.nextInt(available.size()));
                    schedule.get(day).get(shift).add(pick);
                    daysWorked.put(pick, daysWorked.getOrDefault(pick, 0) + 1);
                }
            }
        }
    }

    private boolean alreadyAssigned(String day, String emp) {
        for (String shift : SHIFTS) {
            if (schedule.get(day).get(shift).contains(emp)) return true;
        }
        return false;
    }

    public void printSchedule() {
        for (String day : DAYS) {
            System.out.println("\n" + day + ":");
            for (String shift : SHIFTS) {
                List<String> emps = schedule.get(day).get(shift);
                System.out.println("  " + shift + ": " + (emps.isEmpty() ? "Employees Unavailable" : String.join(", ", emps)));
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Scheduler scheduler = new Scheduler();
        List<String> employees = new ArrayList<>();
        System.out.print("Enter number of employees: ");
        int n = Integer.parseInt(scanner.nextLine().trim());
        for (int i = 0; i < n; i++) {
            String emp;
            while (true) {
                System.out.print("Enter name of employee #" + (i+1) + ": ");
                emp = scanner.nextLine().trim();
                if (emp.isEmpty() || employees.contains(emp)) {
                    System.out.println("Invalid or duplicate name. Please enter a unique employee name.");
                } else {
                    break;
                }
            }
            employees.add(emp);
        }
        System.out.println("\nEnter preferences (leave blank to skip):");
        for (String emp : employees) {
            System.out.print("Does " + emp + " have any shift preferences? (y/n): ");
            String hasPref = scanner.nextLine().trim().toLowerCase();
            if (hasPref.equals("y")) {
                for (String day : DAYS) {
                    for (String shift : SHIFTS) {
                        System.out.print("Does " + emp + " prefer " + shift + " on " + day + "? (y/n, blank to skip): ");
                        String pref = scanner.nextLine().trim().toLowerCase();
                        if (pref.equals("y")) {
                            scheduler.addEmployeePreference(emp, day, shift);
                        }
                    }
                }
            }
        }
        for (String emp : employees) {
            scheduler.preferences.putIfAbsent(emp, new HashMap<>());
        }
        scheduler.assignShifts();
        scheduler.printSchedule();
    }
}

File: python_implementation/scheduler.py
import random
from collections import defaultdict

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["morning", "afternoon", "evening"]

class Scheduler:
    def __init__(self):
        self.preferences = defaultdict(lambda: defaultdict(list))
        self.schedule = defaultdict(lambda: defaultdict(list))
        self.days_worked = defaultdict(int)

    def add_employee_preference(self, name, day, shift):
        if day in DAYS and shift in SHIFTS:
            self.preferences[name][day].append(shift)

    def assign_shifts(self):
        for day in DAYS:
            for shift in SHIFTS:
                candidates = [name for name, prefs in self.preferences.items()
                              if day in prefs and shift in prefs[day]
                              and self.days_worked[name] < 5
                              and name not in sum(self.schedule[day].values(), [])]
                random.shuffle(candidates)
                assigned = candidates[:2]

                for emp in assigned:
                    self.schedule[day][shift].append(emp)
                    self.days_worked[emp] += 1

                while len(self.schedule[day][shift]) < 2:
                    available = [name for name in self.preferences.keys()
                                 if self.days_worked[name] < 5
                                 and name not in sum(self.schedule[day].values(), [])]
                    if not available:
                        break
                    choice = random.choice(available)
                    self.schedule[day][shift].append(choice)
                    self.days_worked[choice] += 1

    def print_schedule(self):
        for day in DAYS:
            print(f"\n{day}:")
            for shift in SHIFTS:
                emps = self.schedule[day][shift]
                if len(emps) < 2:
                    print(f"  {shift.capitalize()}: Employees unavailable")
                else:
                    print(f"  {shift.capitalize()}: {', '.join(emps)}")

# Run the below code to use Scheduler via CLI

if __name__ == "__main__":
    scheduler = Scheduler()
    employees = []
    n = int(input("Enter number of employees: "))
    for i in range(n):
        emp = input(f"Enter name of employee #{i+1}: ").strip()
        while not emp or emp in employees:
            print("Invalid or duplicate name. Please enter a unique employee name.")
            emp = input(f"Enter name of employee #{i+1}: ").strip()
        employees.append(emp)
    print("\nEnter preferences (leave blank to skip):")
    for emp in employees:
        has_pref = input(f"Does {emp} have any shift preferences? (y/n): ").strip().lower()
        if has_pref == 'y':
            for day in DAYS:
                for shift in SHIFTS:
                    pref = input(f"Does {emp} prefer {shift} on {day}? (y/n, blank to skip): ").strip().lower()
                    if pref == 'y':
                        scheduler.add_employee_preference(emp, day, shift)
    for emp in employees:
        if emp not in scheduler.preferences:
            scheduler.preferences[emp]
    scheduler.assign_shifts()
    scheduler.print_schedule()
